### PR TITLE
Double propose detection

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #3133

---

implement proposer policing

### CODE REVIEW

File: shared/featureconfig/config.go [hunk 1]
```diff
@@ -119,6 +120,10 @@ func ConfigureBeaconFeatures(ctx *cli.Context) {
 		log.Warn("Enabled unsafe total balance cache")
 		cfg.EnableTotalBalanceCache = true
 	}
+	if ctx.GlobalBool(HashSlingingSlasherFlag.Name) {
+		log.Warn("Data storage requirements can be very high!")
```
Comment (terencechain) [inline shared/featureconfig/config.go]:
```suggestion
		log.Warn("Enabled double propose detection, data storage requirements can be very high!")
```

File: slasher/proposer_slash.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	proposalExists := epochProposalBitlist[epoch].BitAt(validatorID)
+	if !proposalExists {
+		epochProposalBitlist[epoch].SetBitAt(validatorID, true)
+		return true, nil
+	}
+	return false, nil
+}
+
+func insertSort(data []uint64, element uint64) []uint64 {
```
Comment (terencechain) [inline slasher/proposer_slash.go]: Move `insertSort` and `truncateItems` under utils, these are too generalized to be in this file

Comment (shayzluf) [reply slasher/proposer_slash.go]: should i add it to shared/sliceutil ?

Comment (shayzluf) [reply slasher/proposer_slash.go]: added to utils

File: slasher/utils.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Package slasher implements slashing validation
```
Comment (terencechain) [inline slasher/utils.go]: I meant under /shared/sliceutil, it seems general enough that should be shared across codebase

Comment (shayzluf) [reply slasher/utils.go]: great
moving it 
Thanks

File: shared/sliceutil/slice.go [hunk 4]
```diff
@@ -259,3 +263,23 @@ func IntersectionByteSlices(s ...[][]byte) [][]byte {
 	}
 	return inter
 }
+
+// InsertSort inserts an element into a sorted slice at its index.
```
Comment (terencechain) [inline shared/sliceutil/slice.go]: Tests for `InsertSort` and `TruncateItems`?? 😅

Comment (shayzluf) [reply shared/sliceutil/slice.go]: sure, missed it somehow

Comment (shayzluf) [reply shared/sliceutil/slice.go]: thanks
Added

Comment (nisdas) [inline shared/sliceutil/slice.go]:
```suggestion
// InsertSort inserts a uint64 value into a sorted slice at its index.
```

File: shared/sliceutil/slice.go [hunk 5]
```diff
@@ -259,3 +263,23 @@ func IntersectionByteSlices(s ...[][]byte) [][]byte {
...
+func InsertSort(data []uint64, element uint64) []uint64 {
+	index := sort.Search(len(data), func(i int) bool { return data[i] > element })
+	data = append(data, uint64(0))
+	copy(data[index+1:], data[index:])
+	data[index] = element
+	return data
+}
+
+// TruncateItems truncate a sorted list at a minimum item value.
```
Comment (nisdas) [inline shared/sliceutil/slice.go]:
```suggestion
// TruncateItems truncates a sorted list at a minimum uint64 value.
```

File: slasher/detection/proposer_slash.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+)
+
+var epochProposalBitlist map[uint64]bitfield.Bitlist
+var currentEpoch, weakSubjectivityPeriod uint64
+var epochs []uint64
+
+func init() {
+	epochProposalBitlist = make(map[uint64]bitfield.Bitlist)
+	weakSubjectivityPeriod = uint64(54000)
```
Comment (nisdas) [inline slasher/detection/proposer_slash.go]: this should be from our config file not hardcoded over here

File: slasher/detection/proposer_slash.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+// creating a slashable event.
+// returns true if it is the first time this
+// validatorID propose a block in this epoch or not.
+func CheckNewProposal(currentEpoch uint64, epoch uint64, validatorID uint64) (bool, error) {
+	if currentEpoch > weakSubjectivityPeriod && epoch < currentEpoch-weakSubjectivityPeriod {
+		return false, errors.New("epoch is obsolete = before weak subjectivity period")
+	}
+
+	if _, ok := epochProposalBitlist[epoch]; !ok {
+		epochProposalBitlist[epoch] = bitfield.NewBitlist(300000)
```
Comment (nisdas) [inline slasher/detection/proposer_slash.go]: what is this value ? where does it come from `300000`

File: slasher/detection/proposer_slash.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		epochs = sliceutil.InsertSort(epochs, epoch)
+		var truncate bool
+		var itemsToTruncate []uint64
+		if currentEpoch > weakSubjectivityPeriod {
+			truncate, epochs, itemsToTruncate = sliceutil.TruncateItems(epochs, currentEpoch-weakSubjectivityPeriod)
```
Comment (nisdas) [inline slasher/detection/proposer_slash.go]: why are you returning a boolean from this ? the name of the function is misleading in that case

**Comment by rauljordan** [comment]
Hi @shayzluf, any update on this?

**Comment by shayzluf** [comment]
@rauljordan right now it seems that it doesn't serve  any purpose. the bitmap can't detect receiving the same attestation twice or a overlapping indexed attestation. what leads this optimization to produce many false positives. still waiting for more input from the research team

File: slasher/detection/proposer_slash.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+var epochs []uint64
+
+func init() {
+	epochProposalBitlist = make(map[uint64]bitfield.Bitlist)
+	weakSubjectivityPeriod = uint64(54000)
+}
+
+// CheckNewProposal checks weather a new proposal is allowed or
+// creating a slashable event.
+// returns true if it is the first time this
```
Comment (terencechain) [inline slasher/detection/proposer_slash.go]: Can we turn this function comment into a proper paragraph please?

File: slasher/detection/proposer_slash.go (hunk 7 above)
Comment (terencechain) [reply slasher/detection/proposer_slash.go]: +1

File: slasher/detection/proposer_slash.go (hunk 8 above)
Comment (terencechain) [reply slasher/detection/proposer_slash.go]: +1

**Comment by nisdas** [comment]
@shayzluf whats the update on this ?

**Comment by shayzluf** [comment]
it seems that it is as Proto mentioned. we should be easily detect false positives(thinking repeated block headers as double propose) . while reducing

**Comment by nisdas** [comment]
Sounds good, as long as we can flush out the false positives easily

**Comment by shayzluf** [comment]
we should store block header root as keys so it should be easy to flush out fast

File: slasher/detection/proposer_slash.go (hunk 8 above)
Comment (shayzluf) [reply slasher/detection/proposer_slash.go]: fixed, thanks

File: slasher/detection/proposer_slash.go (hunk 6 above)
Comment (shayzluf) [reply slasher/detection/proposer_slash.go]: fixed, thanks

**Pull request closed without merging**